### ISSUE TITLE
compile_boolean_expression: doesn't produce the right result when term ends with multiple parentheses

### ISSUE THREAD

**Issue by exbuddha** [exbuddha/brat#8]
There are cases where compile_boolean_expression: returns a wrong result. One case is when the original term has nested terms and ends with multiple parentheses:

```
$ compile_boolean_expression: c el pl "a < 3.14 && ! ( 1 == 1 || (2 == 1 && ! 3 < 1 ) )"
$ echo "$c"
[[ a  <  3.14  ]] &&  !  ([[  1  ==  1 ]]  ||  ([[ 2  ==  1 ]]  &&  [[ !  3  <  1 ) ) ]]
$
```

Note that the parentheses at the end are placed inside the last double-bracketed term. This should be fixed.

**Comment by exbuddha** [comment]
`locate_expr:()` is most likely problematic.

**Comment by exbuddha** [comment]
Line 1336: `compile_boolean_expression()` is where things go wrong.

`_compiled_expr="${_compiled_expr:0:_start_of_expr}[[ ${_compiled_expr:_start_of_expr:_end_of_expr-_start_of_expr} ]]${_compiled_expr:_end_of_expr}"`

This approach is a little too naive and it causes problems at the end of expressions when there are nested parentheses. The entire function and its sister function `locate_expr:()` have to be code reviewed again so the flaw in the logic can be found.

**Comment by exbuddha** [comment]
As mentioned, one possible solution is to carry `_p_stack` around but... a cold beer sounds better.

**Comment by exbuddha** [comment]
Issues:
  - `$_start_of_expr` needs to be accurately updated in `locate_expr:` (considering blank characters)
  - `_p_stack` needs to be correctly popped all the way through the entire expression and not just while there are leading open parenthesis found during forward lookup. Dequeuing it is also not the right thing to do because...
  - `$_end_of_expr` is used to do backward lookup and to pop `_p_stack` but this is not the right approach since it is not clear what parenthesis is actually closed off.

Basically, most of the problem is in `locate_expr:` because of the naive approach. There might be other small problems in `compile_boolean_expression:`.

**Comment by exbuddha** [comment]
Fix:
  - `compile_boolean_expression:` needs to be redesigned. There is much better approach to this problem.
  - Define `_p_stack` in `compile_boolean_expression:` as mentioned before to keep track of the work done in that function and to correctly locate parentheses between function calls.
  - Once a logical operator is being searched for (in the main while loop inside `compile_boolean_expression:`), update `_p_stack` and use `$_aop_index` to accurately mark the left and right operands. This makes locating the spot where the test brackets need to opened and closed much simpler and it also helps find and bypass extra parentheses around the boolean term that is being processed.
  - Keep accurate track of the predicates as is currently done. Review the logic. It is most likely fine.
